### PR TITLE
Remove early extension system initialization from Greaselion.

### PR DESCRIPTION
**Pull request by goodov**
`GreaselionServiceFactory` performed incorrect extension service initialization which led to broken startup service dependency. `declarative_net_request::RulesMonitorService` should be created before extensions are loaded, but the call in `GreaselionServiceFactory` forced all extensions to be loaded before all `BrowserContextKeyedService`s had a chance to startup.

The correct extension load on startup is done via `ProfileManager::DoFinalInitForServices`. At this point all services are created.

Resolves https://github.com/brave/brave-browser/issues/30854

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/greaselion/browser/greaselion_service_impl.cc [hunk 1]
```diff
@@ -227,7 +227,6 @@ GreaselionServiceImpl::GreaselionServiceImpl(
     : download_service_(download_service),
       install_directory_(install_directory),
       extension_system_(extension_system),
```
Comment (atuchin-m) [inline components/greaselion/browser/greaselion_service_impl.cc:229]: Do we still need to pass `extension_system` in ctor?
maybe replace `SetExtensionService()` to `SetExtensionSystem()`

File: components/greaselion/browser/greaselion_service_impl.cc [hunk 2]
```diff
@@ -254,6 +253,12 @@ void GreaselionServiceImpl::Shutdown() {
                          base::BindOnce(&DeleteExtensionDirs, extension_dirs_));
 }
 
+void GreaselionServiceImpl::SetExtensionService(
+    extensions::ExtensionService* extension_service) {
+  DCHECK(extension_service);
+  extension_service_ = extension_service;
```
Comment (atuchin-m) [inline components/greaselion/browser/greaselion_service_impl.cc:259]: `DCHECK(!extension_service_)` ?

File: test/data/extensions/declarative_net_request/manifest.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
+{
+    "name": "Declarative net request",
+    "version": "0.1",
+    "manifest_version": 3,
+    "background": {},
+    "permissions": ["webRequest", "declarativeNetRequest", "declarativeNetRequestFeedback"],
```
Comment (atuchin-m) [inline test/data/extensions/declarative_net_request/manifest.json:6]: minor: do we need `webRequest` here ?

**Comment by atuchin-m** [comment]
BTW,`components/greaselion/` depends on `extensions/`. I though it is illegal.

**Review by atuchin-m — APPROVED** [review]
LGTM with a few nits

File: test/data/extensions/declarative_net_request/manifest.json (hunk 3 above)
Comment (goodov) [reply test/data/extensions/declarative_net_request/manifest.json]: removed webRequest and declarativeNetRequestFeedback

File: components/greaselion/browser/greaselion_service_impl.cc (hunk 1 above)
Comment (goodov) [reply components/greaselion/browser/greaselion_service_impl.cc:229]: decided to keep changes in this PR to minimum. other improvements are possible here, but can be done as a follow-up/refactoring.

**Review by iefremov — APPROVED** [review]

**Review by SergeyZhukovsky — APPROVED** [review]
android tests ++

**Pull request merged**